gather-scheduler.js

```diff
@@ -18,99 +18,114 @@
-
-      // Gather: setup phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.setup(options)))
-
-      // Enable tracing and network record collection.
-      .then(_ => driver.beginTrace())
-      .then(_ => driver.beginNetworkCollect())
-
-      // Gather: beforePageLoad phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.beforePageLoad(options)))
-
-      // Load the page (if the CLI / extension want it loaded).
-      .then(_ => {
-        if (loadPage) {
-          return driver.gotoURL(url, driver.WAIT_FOR_LOADED);
-        }
-
-        return Promise.resolve();
-      })
+      return driver.forceUpdateServiceWorkers();
+    });
+  }
 
-      // Gather: afterPageLoad phase
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterPageLoad(options)))
+  // Enable tracing and network record collection.
+  static beginPassiveCollection(driver) {
+    return driver.beginTrace().then(_ => {
+      return driver.beginNetworkCollect();
+    });
+  }
 
-      // Disable network collection; grab records.
-      .then(_ => driver.endNetworkCollect())
-      .then(networkRecords => {
-        tracingData.networkRecords = networkRecords;
-      })
+  static endPassiveCollection(driver, gatherers, options, tracingData) {
+    return driver.endNetworkCollect().then(networkRecords => {
+      tracingData.networkRecords = networkRecords;
+    }).then(_ => {
+      return driver.endTrace();
+    }).then(traceContents => {
+      tracingData.traceContents = traceContents;
+    });
+  }
 
-      // Disable tracing; grab records.
-      .then(_ => driver.endTrace())
-      .then(traceContents => {
-        tracingData.traceContents = traceContents;
-      })
+  static _phase(phaseName) {
+    return function(driver, gatherers, options, tracingData) {
+      return GatherScheduler._runPhase(gatherers, gatherer => {
+        return gatherer[phaseName](options, tracingData);
+      });
+    };
+  }
 
-      // Gather: afterTraceCollected phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterTraceCollected(options, tracingData)))
+  static _runPhase(gatherers, gatherFun) {
+    return gatherers.reduce((chain, gatherer) => {
+      return chain.then(_ => gatherFun(gatherer));
+    }, Promise.resolve());
+  }
 
-      // Disconnect the driver.
-      .then(_ => driver.disconnect())
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
+    const self = GatherScheduler;
 
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
+      throw new Error('You must provide a url to scheduler');
+    }
 
+    return GatherScheduler.runSeries([
+      driver.connect.bind(driver),
+      self.setupDriver,
+      self._phase('setup'),
```

Comment:
I'm still not wild about this :) Calling the phase methods with a string accessor seems like a symptom of impedance mismatch in two areas:
- `GatherScheduler` is really a collection of static functions but is calling into real gatherer classes, which means you can't just call `runSeries` over the different phases; you need something for the `this` binding
- Pretending these functions are easily composable because they're all endomorphisms over the same type when they actually aren't. Taking the union of domains is one way to force them in that round hole, but it can easily end with an unstructured mess of arguments being passed in (this would be even more obvious if `options` and `tracingData` were split into their individual properties before being passed into `runSeries`). And because our functions aren't side-effect free, you lose information at the call site of what a function is changing, relying on having to inspect the function itself.

In both cases you trade some clarity and static verifiability for an easier to read list of steps. For purity's sake we'd want to make them actually endomorphisms/monadic, instead of having the state repeatedly reinjected, or if it's the readability we're after we actually want `async`/`await` here. The first approach is probably overkill, the second isn't possible yet.

I personally feel that something closer to the existing `then`-heavy structure, with the driver steps pulled out as you have here, is a better expression of what we're doing here while we're waiting on `await` support to drop, rather than making a pipeline over (conceptually-)orthogonal domains.